Add DEM to the InSAR L1 products

xhuang-jpl opened the pull request:
This PR is to add the DEM in radar grid to all InSAR L1 products (RIFG, RUNW, and ROFF).  The plots of the DEM in radar grid for each L1 product are shown below where the REE 20+5 MHz, 40+5 MHz, and 80 MHz RSLC data are used for the testing.  **The runtime of the running the entire InSAR workflow on dev-5 and dev-7 can be found** [here](https://docs.google.com/spreadsheets/d/1sbM4VpOT2XooknEjY2vCuXp_iTea1QwZ-b5lVl72V-o/edit?gid=0#gid=0)




The runtime pf the prep_hdf5 before and after adding the DEM generation under the dev-5 is listed below,

<img width="508" height="267" alt="image" src="https://github.com/user-attachments/assets/151dfae3-e8a9-4ac5-ae4a-eefd778380b1" />

**RIFG:**

<img width="1189" height="534" alt="image" src="https://github.com/user-attachments/assets/0e00f61f-36b4-40d8-a9a1-2d5a5727abf9" />

**RUNW:**

<img width="1189" height="534" alt="image" src="https://github.com/user-attachments/assets/fc6d7a4e-37fa-4fce-889f-dd0626870477" />


**ROFF:**

<img width="590" height="522" alt="image" src="https://github.com/user-attachments/assets/f3fc4424-8635-458d-bc40-59bf744bb41b" />

Commit:
sm r3.4

Commit:
merge R4.0.2beta-1

Commit:
change the SM commit id for R4.0.2

Commit:
merge upstream develop branch

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge branch 'develop' of https://github-fn.jpl.nasa.gov/isce-3/isce into develop

Commit:
Merge branch 'develop' of https://github-fn.jpl.nasa.gov/isce-3/isce into develop

Commit:
Merge branch 'develop' of https://github-fn.jpl.nasa.gov/isce-3/isce into develop

Commit:
Merge branch 'develop' of https://github-fn.jpl.nasa.gov/isce-3/isce into develop

Commit:
Merge branch 'develop' of https://github-fn.jpl.nasa.gov/isce-3/isce into develop

Commit:
Merge branch 'develop' of https://github-fn.jpl.nasa.gov/isce-3/isce into develop

Commit:
Merge branch 'develop' of https://github-fn.jpl.nasa.gov/isce-3/isce into develop

Commit:
add the dem generation placeholder function

Commit:
add the dem layer

Commit:
run the isort

Commit:
add the thresholds

Commit:
add print

Commit:
use topo for the dem in rdr generation

Commit:
add block processing

Commit:
fix bugs

Commit:
isort the InSAR_L1_writer.py

Commit:
remove reduntant

Commit:
remove .vscode folder

Commit:
fix the parameters

Commit:
fix the comments

Commit:
Merge branch 'develop' of https://github-fn.jpl.nasa.gov/isce-3/isce into add_rd_dem

vbrancat commented:
@xhuang-jpl can you please clarify how you did obtain the runtime reported in your PR description?
We have two versions of the DEM dataset in radar coordinate, one for the interferogram at 30 m and one for the offsets at 90 m. Please, benchmark the runtime for all the NISAR modes (e.g., 20+5, 40+5 and 80) and for running rdr2geo on a radar grid of 30 m and 90 m. Thanks

vbrancat commented on python/packages/nisar/products/insar/InSAR_L1_writer.py:
```diff
@@ -447,12 +451,41 @@ def add_pixel_offsets_to_swaths_group(self):
-            for attr in ['mean_value', 'min_value',
-                         'max_value', 'sample_stddev']:
-                offset_group['digitalElevationModel'].attrs[attr] = 0.0
+            if self.dem_file is not None:
+                threshold = self.cfg['processing']['rdr2geo']['threshold']
+                numiter = self.cfg['processing']['rdr2geo']['numiter']
+                extraiter = self.cfg['processing']['rdr2geo']['extraiter']
+                lines_per_block = self.cfg['processing']['rdr2geo']['lines_per_block']
+
+                scratch_path = pathlib.Path(self.cfg['product_path_group']['scratch_path'])
+                # create seperate directory within scratch dir for rdr2geo run
+                rdr2geo_scratch_path = scratch_path / 'rdr2geo' / f'freq{freq}'
+                rdr2geo_scratch_path.mkdir(parents=True, exist_ok=True)
+                out_dem_rdr_path = rdr2geo_scratch_path / f'{self.product_info.ProductType}_offsets_dem.rdr'
+
+                # check if gpu ok to use
+                use_gpu = isce3.core.gpu_check.use_gpu(self.cfg['worker']['gpu_enabled'],
+                                                       self.cfg['worker']['gpu_id'])
+                if use_gpu:
+                    # Set the current CUDA device.
+                    device = isce3.cuda.core.Device(self.cfg['worker']['gpu_id'])
+                    isce3.cuda.core.set_device(device)
+
```
There is a lot of redundancy here. You should separate this code into its appropriate function since you will be reusing it to generate the DEM on the radar grid at 30 m and 90 m

xhuang-jpl replied:
Both RIFG and RUNW share the same code section above, and there is no code duplication  for the RIFG and RUNW

vbrancat replied:
what about ROFF? we need to generate the DEM also for the offsets.

vbrancat commented on python/packages/nisar/products/insar/InSAR_L1_writer.py:
```diff
@@ -447,12 +451,41 @@ def add_pixel_offsets_to_swaths_group(self):
-            for attr in ['mean_value', 'min_value',
-                         'max_value', 'sample_stddev']:
-                offset_group['digitalElevationModel'].attrs[attr] = 0.0
+            if self.dem_file is not None:
```
This if/else statement is useless since for the InSAR workflow, the DEM is a required entry of the schema. See [here](https://github.com/isce-framework/isce3/blob/c799f50a6134106ce9e7073a4da781777b3193a8/share/nisar/schemas/insar.yaml#L26). Please, remove the if/else

xhuang-jpl replied:
thanks @vbrancat , I think there is no harm to have the code deal with more cases.

vbrancat commented on python/packages/nisar/products/insar/InSAR_L1_writer.py:
```diff
@@ -630,12 +663,41 @@ def add_interferogram_to_swaths_group(self):
-            for attr in ['mean_value', 'min_value',
-                         'max_value', 'sample_stddev']:
-                igram_group['digitalElevationModel'].attrs[attr] = 0.0
+            if self.dem_file is not None:
+                threshold = self.cfg['processing']['rdr2geo']['threshold']
+                numiter = self.cfg['processing']['rdr2geo']['numiter']
+                extraiter = self.cfg['processing']['rdr2geo']['extraiter']
+                lines_per_block = self.cfg['processing']['rdr2geo']['lines_per_block']
+
+                scratch_path = pathlib.Path(self.cfg['product_path_group']['scratch_path'])
+                # create seperate directory within scratch dir for rdr2geo run
+                rdr2geo_scratch_path = scratch_path / 'rdr2geo' / f'freq{freq}'
+                rdr2geo_scratch_path.mkdir(parents=True, exist_ok=True)
+                out_dem_rdr_path = rdr2geo_scratch_path / f'{self.product_info.ProductType}_ifgram_dem.rdr'
```
Same here. A lot of redundant line of codes which can be more nicely organized into a stand alone function

xhuang-jpl replied:
those codes are to prepare the input parameters for the `generate_dem_rdr` function.

hfattahi commented on python/packages/nisar/products/insar/utils.py:
```diff
@@ -370,6 +372,128 @@ def _compute_subswath_mask_id(azi_idx,
+    if dem_interp_method == 'BICUBIC':
+        interp_method = isce3.core.DataInterpMethod.BICUBIC
+    if dem_interp_method == 'NEAREST':
+        interp_method = isce3.core.DataInterpMethod.NEAREST
```
Looks like if the user of this function passes anything beyond "BILINEAR", "BICUBIC" and "NEAREST", the interp_method is set to BIQUINTIC. Even if user passes dem_interp_method = 'ANYTHING unrelated', it will still silently ignored and set to BIQUINTIC. I don't think if that is a correct begavior. We should let the underlaying functions throw an exception if the interolation method is not valid. I suggest the following minor change:

```suggestion
    
   try:
         interp_method = getattr(isce3.core.DataInterpMethod, dem_interp_method)
    except AttributeError:
         raise ValueError(f"invalid interpolation method: {dem_interp_method}")
```

hfattahi commented on python/packages/nisar/products/insar/utils.py:
```diff
@@ -370,6 +372,128 @@ def _compute_subswath_mask_id(azi_idx,
+
+    Parameters
+    ---------
+    radar_grid_obj : SLC
```
```suggestion
    radar_grid_obj : isce3.product.RadarGridParameters
```

hfattahi commented on python/packages/nisar/products/insar/utils.py:
```diff
@@ -370,6 +372,128 @@ def _compute_subswath_mask_id(azi_idx,
+    orbit_obj : isce3.core.Orbit
+        The SLC object for the secondary RSLC
+    dem_file  : str
+        DEM file
```
```suggestion
        Input DEM file in geocoded coordinates
```

hfattahi commented on python/packages/nisar/products/insar/utils.py:
```diff
@@ -370,6 +372,128 @@ def _compute_subswath_mask_id(azi_idx,
+    out_dem_rdr_path : str
+        output path of the DEM in radar grid
+    use_gpu : boolean (default: True)
+        Indicator to use the GPU or not
```
```suggestion
        Indicator to use GPU for rdr2geo computations
```

hfattahi commented on python/packages/nisar/products/insar/utils.py:
```diff
@@ -370,6 +372,128 @@ def _compute_subswath_mask_id(azi_idx,
+    extraiter : integer (default: 10)
+         Maximum number of secondary iterations
+    lines_per_block : integer (default: 1000)
+         Lines per block to write the generated DEM to the hard drive
```
```suggestion
         Lines per block to run rdr2geo
```

hfattahi reviewed (commented):
Thanks @xhuang-jpl for this addition which addresses [this request](https://github.com/isce-framework/nisar-workflows/issues/5). Here are few minor comments. I agree with @vbrancat that we need to document the run time increase for this addition.

hfattahi commented on python/packages/nisar/products/insar/utils.py:
```diff
@@ -370,6 +372,128 @@ def _compute_subswath_mask_id(azi_idx,
+
+    x_raster, y_raster, incidence_raster,\
+        heading_raster, local_incidence_raster, local_psi_raster,\
+            simulated_amplitude_raster, shadow_raster = [None] * 8
+    rdr2geo_obj.topo(dem_raster, x_raster, y_raster, dem_src,
+                     incidence_raster, heading_raster, local_incidence_raster,
+                     local_psi_raster, simulated_amplitude_raster,
+                     shadow_raster, None, None)
```
Not a big deal but for clarity which rasters are you ignoring. 
```suggestion
            simulated_amplitude_raster, shadow_raster, ground_to_sat_x_ratser, ground_to_sat_y_raster = [None] * 10
    rdr2geo_obj.topo(dem_raster, x_raster, y_raster, dem_src,
                     incidence_raster, heading_raster, local_incidence_raster,
                     local_psi_raster, simulated_amplitude_raster,
                     shadow_raster, ground_to_sat_x_ratser, ground_to_sat_y_raster )
```

vbrancat commented:
@xhuang-jpl, please address comments on this PR and provide evidence that there is a marginal increase of runtime

Commit:
Merge branch 'develop' of https://github.com/isce-framework/isce3 into add_rd_dem

Commit:
address Heresh's comments

Commit:
Merge branch 'add_rd_dem' of github.com:xhuang-jpl/isce3 into add_rd_dem

xhuang-jpl commented:
Thanks @vbrancat and @hfattahi  I have addressed your comments, and benchmarked the runtime on the dev-5 using the 20+5, 40+5, and 80 MHz data. In average, the runtime is increased by about 30 seconds for the RIFG, and 10 seconds for the RUNW, and 5 seconds for the ROFF.

vbrancat commented:
@xhuang-jpl I am still not convinced about your runtimes. My understanding is that you have benchmarked everything on a GPU. However, we are going to produce RIFG and RUNW on a c5.4xlarge. Therefore, you will have to benchmark RIFG and RUNW on dv-7. Also, can you comment on the posting of the DEM for the interferogram and the offsets? Can you point me at the runconfigs you have used

Commit:
record the prepare_insar_hdf5 runtime

hfattahi reviewed (approved):
Thanks @xhuang-jpl for addressing my comments. Glad to see the run time on both CPU and GPU has not increased too much. LGTM.

vbrancat reviewed (approved):
LGTM, thanks!
